go/libkb/wot.go

```diff
@@ -198,48 +202,66 @@ func fetchWot(mctx MetaContext, username *string) (res []serverWotVouch, err err
 	return response.Vouches, nil
 }
 
-func FetchMyWot(mctx MetaContext) (res []keybase1.WotVouch, err error) {
-	defer mctx.Trace("FetchMyWot", func() error { return err })()
-	serverVouches, err := fetchWot(mctx, nil)
+// fetchUserWotByVoucher gets vouches written for vouchee. If voucher is
+// specified, only get the vouch (of any status) written by voucher for vouchee.
+// If voucheeUser has been previously loaded, pass it in to prevent reloading
+// (there is no harm in reloading).
+func fetchUserWotByVoucher(mctx MetaContext, vouchee string, voucher *string, voucheeUser *User) (res []keybase1.WotVouch, err error) {
+	vouches, err := fetchWot(mctx, &vouchee, voucher)
 	if err != nil {
-		mctx.Debug("error fetching pending web-of-trust vouches: %s", err.Error())
+		mctx.Debug("error fetching web-of-trust vouches for %s: %s", vouchee, err.Error())
 		return nil, err
 	}
-	me, err := LoadMe(NewLoadUserArgWithMetaContext(mctx))
-	if err != nil {
-		return nil, fmt.Errorf("error loading myself: %s", err.Error())
+	if voucheeUser == nil {
+		voucheeUser, err = LoadUser(NewLoadUserArgWithMetaContext(mctx).WithName(vouchee))
+		if err != nil {
+			return nil, fmt.Errorf("error loading vouchee: %s", err.Error())
+		}
 	}
-	for _, serverVouch := range serverVouches {
-		vouch, err := transformUserVouch(mctx, serverVouch, me)
+	for _, serverVouch := range vouches {
+		vouch, err := transformUserVouch(mctx, serverVouch, voucheeUser)
 		if err != nil {
-			mctx.Debug("error validating server-reported pending web-of-trust vouches: %s", err.Error())
+			mctx.Debug("error validating server-reported web-of-trust vouches for %s: %s", vouchee, err.Error())
 			return nil, err
 		}
 		res = append(res, vouch)
 	}
-	mctx.Debug("found %d web-of-trust vouches", len(res))
+	mctx.Debug("found %d web-of-trust vouches for %s", vouchee, len(res))
 	return res, nil
 }
 
-func FetchUserWot(mctx MetaContext, username string) (res []keybase1.WotVouch, err error) {
-	defer mctx.Trace("FetchUserWot", func() error { return err })()
-	vouches, err := fetchWot(mctx, &username)
+// FetchMyWot gets all vouches (of any status) written for me
+func FetchMyWot(mctx MetaContext) (res []keybase1.WotVouch, err error) {
+	defer mctx.Trace("FetchMyWot", func() error { return err })()
+	me, err := LoadMe(NewLoadUserArgWithMetaContext(mctx))
 	if err != nil {
-		mctx.Debug("error fetching web-of-trust vouches for %s: %s", username, err.Error())
-		return nil, err
+		return nil, fmt.Errorf("error loading myself: %s", err.Error())
 	}
-	vouchee, err := LoadUser(NewLoadUserArgWithMetaContext(mctx).WithName(username))
+	myName := me.GetName()
+	return fetchUserWotByVoucher(mctx, myName, nil, me)
+}
+
+// FetchMyWotByUser gets vouch (of any status) written for me by the voucher
+func FetchMyWotByUser(mctx MetaContext, voucher string) (res []keybase1.WotVouch, err error) {
```

Comment:
added this "ByUser" method for completeness, but maybe we don't need it

Reply:
this is a lot of publicly exposed methods with really similar names that take really similar parameters and do subtly different things. i suggest we consider collapsing all of these together into one method that takes an argument type instead of multiple methods. two methods was one thing, but four is... a little overwhelming in my opinion. and an argument type will allow us to more-easily add behavior later if necessary. so something like:
```go
// defined in the avdl
type FetchVouchesArg {
    voucher *string
    vouchee *string
}

func FetchVouches(mctx MetaContext, arg keybase1.FetchVouchesArg) {
    // any necessary input validation like... 
    // at least one of voucher / vouchee is set
    // maybe load the user(s), and then kick out an error here if its not a valid user
    // if voucher is me, get all statuses
    // if vouchee is me, get all statuses
    // etc.
}
```
correct me if i'm wrong, i think if from here, we expect the caller to specify (i.e. no friendly defaults) then the only parameters that are necessary to determine intent is voucher and vouchee. which seems like the right tradeoff to me.

Reply:
let me know what you think of the new `FetchUserWot`